packages/react-data-grid/src/ColumnMetrics.ts

```diff
@@ -0,0 +1,133 @@
+    if (column.width) continue;
+
+    if (unallocatedWidth <= 0) {
+      column.width = minColumnWidth;
+    }
+
+    column.width = columnWidth < minColumnWidth ? minColumnWidth : columnWidth;
+  }
+}
+
+function setColumnOffsets(columns: Column[]): void {
+  let left = 0;
+  for (const column of columns) {
+    column.left = left;
+    left += column.width;
+  }
+}
+
+const getTotalColumnWidth = (columns: Column[]): number => columns.reduce((acc, c) => acc + (c.width || 0), 0);
+
+export function recalculate(metrics: Metrics) {
+  // clone columns so we can safely edit them:
+  let columns = cloneColumns(metrics.columns);
+  // compute width for columns which specify width
+  setColumnWidths(columns, metrics.totalWidth);
+
+  const width = getTotalColumnWidth(columns);
+  const unallocatedWidth = metrics.totalWidth - width - getScrollbarSize();
+
+  // compute width for columns which doesn't specify width
+  setDefferedColumnWidths(columns, unallocatedWidth, metrics.minColumnWidth);
+
+  // compute left offset
+  setColumnOffsets(columns);
+
+  const frozenColumns = columns.filter(c => isFrozen(c));
+  const nonFrozenColumns = columns.filter(c => !isFrozen(c));
+  columns = frozenColumns.concat(nonFrozenColumns);
+  columns.forEach((c, i) => c.idx = i);
+  return {
+    columns,
+    width,
+    totalWidth: metrics.totalWidth,
+    totalColumnWidth: getTotalColumnWidth(columns),
+    minColumnWidth: metrics.minColumnWidth
+  };
+}
+
+/**
+ * Update column metrics calculation by resizing a column.
+ */
+export function resizeColumn(metrics: Metrics, index: number, width: number) {
+  const column = getColumn(metrics.columns, index);
+  const metricsClone = { ...metrics };
+  metricsClone.columns = cloneColumns(metrics.columns);
+
+  const updatedColumn = { ...column };
+  updatedColumn.width = Math.max(width, metrics.minColumnWidth);
+
+  return recalculate(spliceColumn(metricsClone, index, updatedColumn));
+}
+
+type ColumnComparer = (colA: Column, colB: Column) => boolean;
+
+function compareEachColumn(prevColumns: Column[], nextColumns: Column[], isSameColumn: ColumnComparer): boolean {
+  if (getSize(prevColumns) !== getSize(nextColumns)) return false;
+
+  const keys = new Set<string>();
+  const prevColumnsMap = new Map<string, Column>();
+  const nextColumnsMap = new Map<string, Column>();
+
+  for (const column of prevColumns) {
+    keys.add(column.key);
+    prevColumnsMap.set(column.key, column);
+  }
+
+  for (const column of nextColumns) {
+    keys.add(column.key);
+    nextColumnsMap.set(column.key, column);
+  }
+
+  if (keys.size > prevColumnsMap.size) return false;
+
+  // We shouldn't need Array.from(), but TS complains
+  for (const key of Array.from(keys)) {
+    if (!prevColumnsMap.has(key) || !nextColumnsMap.has(key)) return false;
+    const prevColumn = prevColumnsMap.get(key)!;
+    const nextColumn = nextColumnsMap.get(key)!;
+    if (isSameColumn(prevColumn, nextColumn)) return false;
```

Comment:
should it not be `!isSameColumn(prevColumn, nextColumn)`?

Reply:
You're right, fixed it now.